Readme.pod
=pod

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Notifications - Frontend Event Notifications


=head1 SYNOPSIS

  # Register the plugin and several engines
  plugin Notifications => {
    Humane => {
      base_class => 'libnotify'
    },
    JSON => 1
  };

  # Add notification messages in controllers
  $c->notify(warn => 'Something went wrong');

  # Render notifications in templates ...
  %= notifications 'humane';

  # ... or in any other responses
  my $json = { text => 'That\'s my response' };
  $c->render(json => $c->notifications(json => $json));


=head1 DESCRIPTION

L<Mojolicious::Plugin::Notifications> supports several engines
to notify users on events. Notifications will survive multiple
redirects and can be served depending on response types.


=head1 METHODS

L<Mojolicious::Plugin::Notifications> inherits all methods
from L<Mojolicious::Plugin> and implements the following new one.

=head2 register

  plugin Notifications => {
    Humane => {
      base_class => 'libnotify'
    },
    HTML => 1
  };

Called when registering the plugin.

Accepts the registration of multiple L<engines|/ENGINES> for notification
responses. Configurations of the engines can be passed as hash
references. If no configuration should be passed, just pass a scalar value.

All parameters can be set either as part of the configuration
file with the key C<Notifications> or on registration
(that can be overwritten by configuration).


=head1 HELPERS

=head2 notify

  $c->notify(error => 'Something went wrong');
  $c->notify(error => { timeout => 4000 } => 'Something went wrong');

Notify the user about an event.
Expects an event type and a message as strings.
In case a notification engine supports further refinements,
these can be passed in a hash reference as a second parameter.
Event types are free and its treatment is up to the engines,
however notifications of the type C<debug> will only be passed in
development mode.


=head2 notifications

  %= notifications 'humane' => [qw/warn error success/];
  %= notifications 'html';

  $c->render(json => $c->notifications(json => {
    text => 'My message'
  }));

Serve notifications to your user based on an engine.
The engine's name has to be passed as the first parameter
and the engine has to be L<registered|/register> in advance.
Notifications won't be invoked in case no notifications are
in the queue and no further engine parameters are passed.
Engine parameters are documented in the respective plugins.

The engine's name will be camelized. If no namespace is given,
the default namespace is C<Mojolicious::Plugin::Notifications>.

In case no engine name is passed to the notifications method,
an L<assets object|Mojolicious::Plugin::Notifications::Assets>
is returned, bundling all registered engines' assets for use
in the L<AssetPack|Mojolicious::Plugin::AssetPack> pipeline.

  # Register Notifications plugin
  app->plugin('Notifications' => {
    Humane => {
      base_class => 'libnotify'
    },
    Alertify => 1
  });

  # Register AssetPack plugin
  app->plugin('AssetPack');

  # Add notification assets to pipeline
  app->asset('myApp.js'  => 'myscripts.coffee', app->notifications->scripts);
  app->asset('myApp.css' => 'mystyles.scss', app->notifications->styles);

  %# In templates embed assets ...
  %= asset 'myApp.js'
  %= asset 'myApp.css'

  %# ... and notifications (without assets)
  %= notifications 'humane', -no_include;

B<The asset helper option is experimental and may change without warnings!>


=head1 ENGINES

L<Mojolicious::Plugin::Notifications> bundles a couple of different
notification engines, but you can
L<easily write your own engine|Mojolicious::Plugin::Notifications::Engine>.


=head2 Bundled engines

The following engines are bundled with this plugin:
L<HTML|Mojolicious::Plugin::Notifications::HTML>,
L<JSON|Mojolicious::Plugin::Notifications::JSON>,
L<Humane.js|Mojolicious::Plugin::Notifications::Humane>, and
L<Alertify.js|Mojolicious::Plugin::Notifications::Alertify>,


=head1 HOOKS

  app->hook(
    before_notifications => sub {
      my ($c, $notes) = @_;
      $c->app->log('Served ' . $notes->size . ' notifications to ' . $c->stash('user'));
    });

This hook is emitted before any notifications are rendered.
The hook passes the current controller object and a L<Mojo::Collection>
object including all notes.
The hook is emitted no matter if notifications are pending.

B<This hook is EXPERIMENTAL!>


=head1 SEE ALSO

If you want to use C<Humane.js> without L<Mojolicious::Plugin::Notifications>,
you should have a look at L<Mojolicious::Plugin::Humane>,
which was the original inspiration for this plugin.

Without my knowledge (due to a lack of research by myself),
L<Mojolicious::Plugin::BootstrapAlerts> already established
a similar mechanism for notifications using Twitter Bootstrap
(not yet supported by this module).
Accidentally the helper names collide - I'm sorry for that!
On the other hands, that makes these modules in most occasions
compatible.


=head1 HINTS

As flash information is stored in the session, notifications may be lost
in case the session expires using C<session(expires =E<gt> 1)>.


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Notifications


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014-2021, L<Nils Diewald|https://www.nils-diewald.de/>.

Part of the code was written at the
L<Mojoconf 2014|http://www.mojoconf.org/mojo2014/> hackathon.

This program is free software, you can redistribute it
and/or modify it under the terms of the Artistic License version 2.0.

=cut


Changes
1.07 2021-08-25
        - support _format parameters in tests.

1.06 2021-02-08
        - Move bugtracker to GitHub.

1.05 2020-03-31
        - Fixed a bug using $c->b() in HTML engine,
          I wasn't able to test.
        - Improved test suite.

1.04 2019-05-14
        - Updated Mojolicious dependency
          (Thanks to Slaven Rezić!).
        - Fixed bug in docs
          (Thanks Mohammad S Anwar!).

1.03 2019-05-13
        - Introduced 'confirmation' types.
        - Added csrf_token to confirmation routes.
        - Notify on error in alertify.
        - Introduced notify_alertify() function.
        - Update of Test::Mojo::Session to be compatible
          with recent version of Mojolicious.

1.02 2018-05-30
        - Introduced notify_html() function.
        - Use Mojo::Collection for notifications.
        - Introduced "before_notifications" hook.
        - Removed deprecated support of 'Mojolicious::Plugin'
          based engines.

1.01 2017-04-27
        - Fixed version to be compliant with perl/cpan
          versioning (Thanks to Slaven Rezić!).

0.12 2017-01-11
        - Updated to Mojolicious 7.13.

0.11 2016-09-10
        - Added test for lower path notifications in sessions.
        - Kwalitee compliant.

0.10 2015-06-18
        - Introduced example for plugins.

0.09 2015-04-19
        - Update dependency on Test::Mojo::Session.

0.08 2015-04-18
        - Update for use with Mojolicious 6.

0.07 2015-01-25
        - Fix Test::Mojo::Session dependency.

0.06 2014-11-04
        - Fix minor documentation error.
        - Remove deprecated use of object-oriented Mojo::JSON.

0.05 2014-09-15
        - Changed tx->resume() method to hook for multiple redirect support.
        - Flash and stash notifications will now be merged.

0.04 2014-09-07
        - Added experimental AssetPack support.
        - Added abstract M::P::N::Engine package.
        - Engines based on Mojolicious::Plugin are now DEPRECATED.
        - Added M::P::N::Assets package.

0.03 2014-06-01
        - Documentation cleanup.
        - Added Alertify engine.

0.02 2014-06-01
        - Initial submission to CPAN.
        - Added <noscript /> tag to the Humane engine.

0.01 2014-05-26
        - Initial submission to GitHub.


Makefile.PL
#!/usr/bin/env perl
use 5.010001;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
  NAME         => 'Mojolicious::Plugin::Notifications',
  VERSION_FROM => 'lib/Mojolicious/Plugin/Notifications.pm',
  ABSTRACT     => 'Event Notifications for your Users',
  AUTHOR       => 'Nils Diewald',
  BUILD_REQUIRES => {
    'Test::More' => 0,
    'Test::Mojo::Session' => 1.04
  },
  PREREQ_PM => {
    'Mojolicious' => 5.30
  },
  LICENSE      => 'artistic_2',
  META_MERGE   => {
    requires  => {
      perl => '5.010001'
    },
    resources => {
      license     => 'http://www.opensource.org/licenses/artistic-license-2.0',
      repository  => 'https://github.com/Akron/Mojolicious-Plugin-Notifications'
    }
  },
  MIN_PERL_VERSION => '5.010001',
  test => {TESTS => 't/*.t'}
);


lib/Mojolicious/Plugin/Notifications/HTML.pm
package Mojolicious::Plugin::Notifications::HTML;
use Mojo::Base 'Mojolicious::Plugin::Notifications::Engine';
use Exporter 'import';
use Mojo::Util qw/xml_escape/;
use Scalar::Util qw/blessed/;

our @EXPORT_OK = ('notify_html');

# TODO:
#   Make the form have a class instead of the button!

# Exportable function
sub notify_html {
  my $c = shift if blessed $_[0] && $_[0]->isa('Mojolicious::Controller');
  my $type = shift;
  my $param = shift if ref $_[0] && ref $_[0] eq 'HASH';
  my $msg = pop;

  my $str = qq{<div class="notify notify-$type">};
  $str .= blessed $msg && $msg->isa('Mojo::ByteStream') ? $msg : xml_escape($msg);

  # Check for confirmation
  if ($param) {

    # Okay path is defined
    if ($param->{ok}) {
      $str .= '<form action="' . $param->{ok} . '" method="post">';
      $str .= $c->csrf_field if $c;
      $str .= '<button class="ok">' . ($param->{ok_label} // 'OK') . '</button>';
      $str .= '</form>';
    };

    # Cancel path is defined
    if ($param->{cancel}) {
      $str .= '<form action="' . $param->{cancel} . '" method="post">';
      $str .= $c->csrf_field if $c;
      $str .= '<button class="cancel">' . ($param->{cancel_label} // 'Cancel') . '</button>';
      $str .= '</form>';
    };
  };
  return $str . "</div>\n";
};


# Notification method
sub notifications {
  my ($self, $c, $notify_array) = @_;

  my $html = '';
  foreach (@$notify_array) {
    $html .= notify_html($c, @{$_});
  };

  return $c->b($html);
};


1;


__END__

=pod

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Notifications::HTML - Event Notifications using HTML


=head1 SYNOPSIS

  # Register the engine
  plugin Notifications => {
    HTML => 1
  };

  # In the template
  %= notifications 'html'


=head1 DESCRIPTION

This plugin is a simple notification engine for HTML.

If it does not suit your needs, you can easily
L<write your own engine|Mojolicious::Plugin::Notifications::Engine>.


=head1 HELPERS

=head2 notify

See the base L<notify|Mojolicious::Plugin::Notifications/notify> helper.


=head2 notifications

  $c->notify(warn => 'wrong');
  $c->notify(success => 'right');

  %= notifications 'html';
  # <div class="notify notify-warn">wrong</div>
  # <div class="notify notify-success">right</div>

Will render each notification using
L<notify_html|Mojolicious::Plugin::Notifications::HTML/notify_html>.


=head1 EXPORTABLE FUNCTIONS

=head2 notify_html

  use Mojolicious::Plugin::Notifications::HTML qw/notify_html/;

  notify_html(warn => 'This is a warning')
  # <div class="notify notify-warn">This is a warning</div>

  notify_html(announce => {
    ok => 'http://example.com/ok',
    ok_label => 'Okay!'
  # }, 'Confirm, please!')
  # <div class="notify notify-announce">
  #   Confirm, please!
  #   <form action="http://example.com/ok" method="post">
  #     <button>Okay!</button>
  #   </form>
  # </div>

Returns the formatted text in a C<E<lt>div /E<gt>> element
with the class C<notify> and the class C<notify-$type>, where C<$type> is
the notification type you passed.
In case an C<ok> parameter is passed, this will add a POST form
for confirmation. In case an C<ok_label> is passed, this will be the label
for the confirmation button.
In case a C<cancel> parameter is passed, this will add a POST form
for cancelation. In case a C<cancel_label> is passed, this will be the label
for the cancelation button.

If the first parameter is a L<Mojolicious::Controller> object,
the button will have a
L<csrf_token|Mojolicious::Plugin::TagHelpers/csrf_token>
parameter to validate.

This is meant to be used by other engines as a fallback.

B<Confirmation is EXPERIMENTAL!>


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Notifications


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014-2018, L<Nils Diewald|http://nils-diewald.de/>.

This program is free software, you can redistribute it
and/or modify it under the terms of the Artistic License version 2.0.

=cut


lib/Mojolicious/Plugin/Notifications/JSON.pm
package Mojolicious::Plugin::Notifications::JSON;
use Mojo::Base 'Mojolicious::Plugin::Notifications::Engine';

# TODO:
#   Instead of 'ok' and 'cancel', probably pass an additional
#   'confirm' route, that will open a form instead.

has key => 'notifications';

# Nothing to register
sub register {
  my ($plugin, $app, $param) = @_;
  $plugin->key($param->{key}) if $param->{key};
};


# Notification method
sub notifications {
  my ($self, $c, $notify_array, $rule, $json, %param) = @_;

  my $key = $param{key} // $self->key;

  return $json unless @$notify_array;

  if (!$json || ref $json) {
    my @msgs;
    foreach (@$notify_array) {

      # Confirmation message
      if (ref $_->[1] && ref $_->[1] eq 'HASH' && ($_->[1]->{ok} || $_->[1]->{cancel})) {
        my $param = $_->[1];

        my $opt = {};

        # Set confirmation path
        if ($param->{ok}) {
          $opt->{$param->{ok_label} // 'ok'} = {
            method => 'POST',
            url => $param->{ok}
          };
        }

        # Set cancelation path
        if ($param->{cancel}) {
          $opt->{$param->{cancel_label} // 'cancel'} = {
            method => 'POST',
            url => $param->{cancel}
          };
        };

        push @msgs, [$_->[0], ''.$_->[-1], $opt];
      }

      # Normal notification
      else {
        push(@msgs, [$_->[0], ''.$_->[-1]]);
      };
    };

    if (!$json) {
      return { $key => \@msgs }
    }

    # Obect is an array
    elsif (index(ref $json, 'ARRAY') >= 0) {
      push(@$json, { $key => \@msgs });
    }

    # Object is a hash
    elsif (index(ref $json, 'HASH') >= 0) {
      my $n = ($json->{$key} //= []);
      push(@$n, @msgs);
    };
  };

  return $json;
};


1;


__END__

=pod

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Notifications::JSON - Event Notifications in JSON


=head1 SYNOPSIS

  # Register the engine
  plugin Notifications => {
    JSON => 1
  };

  # In the controller
  $c->render(json => $c->notifications(json => $json));


=head1 DESCRIPTION

This plugin is a simple notification engine for JSON.

If it does not suit your needs, you can easily
L<write your own engine|Mojolicious::Plugin::Notifications::Engine>.


=head1 METHODS

L<Mojolicious::Plugin::Notifications::JSON> inherits all methods
from L<Mojolicious::Plugin::Notifications::Engine> and implements
the following new one.


=head2 register

  plugin Notifications => {
    JSON => {
      key => 'notes'
    }
  };

Called when registering the main plugin.
All parameters under the key C<JSON> are passed to the registration.

Accepts the following parameters:

=over 4

=item B<key>

Define the attribute name of the notification array.
Defaults to C<notifications>.

=back


=head1 HELPERS

=head2 notify

  $c->notify(warn => 'wrong');
  $c->notify(confirm => { ok => '/ok', cancel => '/cancel' } => 'Please, confirm!');

  $c->render(json => $c->notifications(json => { text => 'html' }));
  # {
  #   "notifications":[
  #     ["warn", "wrong"],
  #     [
  #       "confirm",
  #       "Please, confirm!",
  #       {
  #         "cancel":{
  #           "method":"POST",
  #           "url":"\/cancel"
  #         },
  #         "ok":{
  #           "method":"POST",
  #           "url":"\/ok"
  #         }
  #       }
  #     ]
  #   ],
  #   "text":"example"
  # }


See the base L<notify|Mojolicious::Plugin::Notifications/notify> helper.

In case an C<ok> or C<cancel> parameter is passed,
this will add an object containing confirmation and/or cancellation paths.
In case an C<ok_label> is passed, this will be the key for the
confirmation object.
In case a C<cancel_label> is passed, this will be the key
for the cancelation object.

Confirmation routes for JSON do not support CSRF protection.


B<Confirmation is EXPERIMENTAL!>

=head2 notifications

  $c->render(json => $c->notifications(json => $json));
  $c->render(json => $c->notifications(json => $json, key => 'notes'));

Merge notifications into your JSON response.

In case JSON is an object, it will inject an attribute
that points to an array reference containing the notifications.
If the JSON is an array, an object is appended with one attribute
that points to an array reference containing the notifications.
If the JSON is empty, an object will be created with one attribute
that points to an array reference containing the notifications.

If the JSON is not of one of the descripted types, it's returned
unaltered.

The name of the attribute can either be given on registration or
by passing a parameter for C<key>.
The name defaults to C<notifications>.


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Notifications


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014-2018, L<Nils Diewald|http://nils-diewald.de/>.

This program is free software, you can redistribute it
and/or modify it under the terms of the Artistic License version 2.0.

=cut


lib/Mojolicious/Plugin/Notifications/Assets.pm
package Mojolicious::Plugin::Notifications::Assets;
use Mojo::Base -strict;

# Don't stringify, in case someone forgot to define an engine
use overload '""' => sub { '' }, fallback => 1;


# Constructor
sub new {
  bless {
    styles => [],
    scripts => []
  }, shift;
};


# Get or add styles
sub styles {
  my $self = shift;
  return sort @{ $self->{styles} } unless @_;
  push(@{$self->{styles}}, @_);
};


# Get or add scripts
sub scripts {
  my $self = shift;
  return sort @{ $self->{scripts} } unless @_;
  push(@{$self->{scripts}}, @_);
};


1;


__END__

=pod

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Notifications::Assets - Collect Assets of Notification Engines


=head1 SYNOPSIS

  use Mojolicious::Plugin::Notifications::Assets;

  my $assets = Mojolicious::Plugin::Notifications::Assets->new;
  $assets->scripts('/myscripts.js');
  $assets->styles('/mystyles.css');


=head1 DESCRIPTION

L<Mojolicious::Plugin::Notifications::Assets> establishes a simple
collector object for assets, used by L<Mojolicious::Plugin::Notifications>.


=head1 METHODS

=head2 new

  my $assets = Mojolicious::Plugin::Notifications::Assets->new;

Create a new assets object.


=head2 scripts

  $assets->scripts('/myscripts.js');
  my @scripts = $assets->scripts;

Add scripts to the asset list or return the collected scripts in sorted order.


=head2 styles

  $assets->styles('/mystyles.css');
  my @styles = $assets->styles;

Add styles to the asset list or return the collected styles in sorted order.


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Notifications


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014, L<Nils Diewald|http://nils-diewald.de/>.

This program is free software, you can redistribute it
and/or modify it under the terms of the Artistic License version 2.0.

=cut


lib/Mojolicious/Plugin/Notifications/Engine.pm
package Mojolicious::Plugin::Notifications::Engine;
use Mojo::Base 'Mojolicious::Plugin';
use Scalar::Util qw/blessed/;

# Register the plugin - optional
sub register {
  # Nothing to register - but has to be called
};


# scripts attribute
sub scripts {
  $_[0]->{scripts} //= [];
  return @{$_[0]->{scripts}} if @_ == 1;
  push(@{shift->{scripts}}, @_);
};


# styles atttribute
sub styles {
  $_[0]->{styles} //= [];
  return @{$_[0]->{styles}} if @_ == 1;
  push(@{shift->{styles}}, @_);
};


# notifications method
sub notifications {
  my $self = shift;
  state $msg = 'No notification engine specified';
  $self->app->log->error($msg . ' for ' . blessed $self);
  return $msg;
};


1;


__END__

=pod

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Notifications::Engine - Abstract Class for Notification Engines


=head1 SYNOPSIS

  package Mojolicious::Plugin::Notifications::MyEngine;
  use Mojo::Base 'Mojolicious::Plugin::Notifications::Engine';

  # Define notifications helper
  sub notifications {
    my ($self, $c, $notifications) = @_;

    my $string = '';
    foreach (@$notifications) {
      $string .= '<blink class="' . $_->[0] . '">' . $_->[-1] . '</blink>';
    };
    return $c->b($string);
  };

=head1 DESCRIPTION

L<Mojolicious::Plugin::Notifications::Engine> is an abstract class
for creating notification engines. It is meant to be used as the base
of notification engine classes.


=head1 METHODS

L<Mojolicious::Plugin::Notifications::Engine> inherits all methods
from L<Mojolicious::Plugin> and implements the following new ones.


=head2 register

  sub register {
    my ($self, $app, $param) = @_;
    # ...
  };

Called when the engine is L<registered|Mojolicious::Plugin::Notifications/register>.
This by default does nothing, but the engine may define assets, helpers, hooks etc.
overriding this method.
The optional parameter will be passed as defined in the registration.


=head2 scripts

  $self->scripts('/mybasescript.js', '/myscript.js');
  print $self->scripts;

Add further script assets, to be used by the
L<scripts|Mojolicious::Plugin::Notifications::Assets/scripts> helper.


=head2 styles

  $self->styles('/mystyles.css', '/mycolors.css');
  print $self->styles;

Add further style assets, to be used by the
L<styles|Mojolicious::Plugin::Notifications::Assets/styles> helper.


=head2 notifications

  # Define notifications method
  sub notifications {
    my ($self, $c, $notifications) = @_;

    my $string = '';
    foreach (@$notifications) {
      $string .= '<blink class="' . $_->[0] . '">' . $_->[-1] . '</blink>';
    };
    return $c->b($string);
  };

Create a notification method.

The C<notifications> method will be called whenever notifications are rendered.
The first parameter passed is the plugin object, the second parameter is the current
controller object and the third parameter is an array reference containing all
notifications as array references.

The first element of the notification is the
notification type, the last element is the message. An optional second element may
contain further parameters in a hash reference.

  %= notifications 'MyEngine', -no_include

Possible flags (boolean parameters marked with a dash) are passed as a hash reference.
All other parameters passed to the L<notifications> helper are simply appended.

The L<bundled engines|/Bundled engines> can serve as good examples on how
to write an engine, especially the simple
L<HTML|Mojolicious::Plugin::Notifications::HTML> engine.


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Notifications


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014, L<Nils Diewald|http://nils-diewald.de/>.

This program is free software, you can redistribute it
and/or modify it under the terms of the Artistic License version 2.0.

=cut


lib/Mojolicious/Plugin/Notifications.pm
package Mojolicious::Plugin::Notifications;
use Mojo::Base 'Mojolicious::Plugin';
use Mojo::Collection qw/c/;
use Mojolicious::Plugin::Notifications::Assets;
use Mojo::Util qw/camelize/;
use Scalar::Util qw/blessed/;

our $TYPE_RE = qr/^[-a-zA-Z_]+$/;

our $VERSION = '1.03';

# TODO:
#   Maybe revert to tx-handler and use session instead of flash!
# TODO:
#   Support Multiple Times Loading
# TODO:
#   Subroutines for Engines should be given directly
# TODO:
#   Engines may use hooks in addition

# Register plugin
sub register {
  my ($plugin, $app, $param) = @_;

  $param ||= {};

  my $debug = $app->mode eq 'development' ? 1 : 0;

  # Load parameter from Config file
  if (my $config_param = $app->config('Notifications')) {
    $param = { %$param, %$config_param };
  };

  $param->{HTML} = 1 unless keys %$param;

  # Add engines from configuration
  my %engine;
  foreach my $name (keys %$param) {
    my $engine = camelize $name;
    if (index($engine,'::') < 0) {
      $engine = __PACKAGE__ . '::' . $engine;
    };

    # Load engine
    my $e = $app->plugins->load_plugin($engine);
    $e->register($app, ref $param->{$name} ? $param->{$name} : undef);
    $engine{lc $name} = $e;
  };

  # Create asset object
  my $asset = Mojolicious::Plugin::Notifications::Assets->new;

  # Set assets
  foreach (values %engine) {
    $asset->styles($_->styles)   if $_->can('styles');
    $asset->scripts($_->scripts) if $_->can('scripts');
  };


  # Add notifications
  $app->helper(
    notify => sub {
      my $c = shift;
      my $type = shift;
      my @msg = @_;

      # Ignore debug messages in production
      return if $type !~ $TYPE_RE || (!$debug && $type eq 'debug');

      my $notes;

      # Notifications already set
      if ($notes = $c->stash('notify.array')) {
        push @$notes, [$type => @msg];
      }

      # New notifications as a Mojo::Collection
      else {
        $c->stash('notify.array' => c([$type => @msg]));
      };
    }
  );


  # Embed notification display
  $app->helper(
    notifications => sub {
      my $c = shift;

      return $asset unless @_;

      my $e_type = lc shift;
      my @param = @_;

      # Get stash notifications
      my $notes = ($c->stash->{'notify.array'} //= c());

      # Get flash notifications
      my $flash = $c->flash('n!.a');

      if ($flash && ref $flash eq 'ARRAY') {

        # Ensure that no harmful types are injected
        unshift @$notes, grep { $_->[0] =~ $TYPE_RE } @$flash;
      };

      # Emit hook for further flexibility
      $app->plugins->emit_hook(
        before_notifications => $c, $notes
      );

      # Delete from stash
      delete $c->stash->{'notify.array'};

      # Nothing to do
      return '' unless $notes->size || @_;

      # Forward messages to notification center
      if (exists $engine{$e_type}) {

        my %rule;
        while ($param[-1] && index($param[-1], '-') == 0) {
          $rule{lc(substr(pop @param, 1))} = 1;
        };

        return $engine{$e_type}->notifications($c, $notes, \%rule, @param);
      }
      else {
        $c->app->log->error(qq{Unknown notification engine "$e_type"});
        return;
      };
    }
  );


  # Add hook for redirects
  $app->hook(
    after_dispatch => sub {
      my $c = shift;

      if ($c->res->is_redirect) {
        # Get notify array from stash
        my $notes = delete $c->stash->{'notify.array'};

        # Check flash notes
        if ($c->flash('n!.a')) {

          # Merge notes with flash
          if ($notes) {
            unshift @$notes, @{$c->flash('n!.a')};
          }

          # Get notes from flash
          else {
            $notes = $c->flash('n!.a');
          };
        };

        if ($notes) {
          $c->flash('n!.a' => $notes);
        };
      };
    });
};


1;


__END__

=pod

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Notifications - Frontend Event Notifications


=head1 SYNOPSIS

  # Register the plugin and several engines
  plugin Notifications => {
    Humane => {
      base_class => 'libnotify'
    },
    JSON => 1
  };

  # Add notification messages in controllers
  $c->notify(warn => 'Something went wrong');

  # Render notifications in templates ...
  %= notifications 'humane';

  # ... or in any other responses
  my $json = { text => 'That\'s my response' };
  $c->render(json => $c->notifications(json => $json));


=head1 DESCRIPTION

L<Mojolicious::Plugin::Notifications> supports several engines
to notify users on events. Notifications will survive multiple
redirects and can be served depending on response types.


=head1 METHODS

L<Mojolicious::Plugin::Notifications> inherits all methods
from L<Mojolicious::Plugin> and implements the following new one.

=head2 register

  plugin Notifications => {
    Humane => {
      base_class => 'libnotify'
    },
    HTML => 1
  };

Called when registering the plugin.

Accepts the registration of multiple L<engines|/ENGINES> for notification
responses. Configurations of the engines can be passed as hash
references. If no configuration should be passed, just pass a scalar value.

All parameters can be set either as part of the configuration
file with the key C<Notifications> or on registration
(that can be overwritten by configuration).


=head1 HELPERS

=head2 notify

  $c->notify(error => 'Something went wrong');
  $c->notify(error => { timeout => 4000 } => 'Something went wrong');

Notify the user about an event.
Expects an event type and a message as strings.
In case a notification engine supports further refinements,
these can be passed in a hash reference as a second parameter.
Event types are free and its treatment is up to the engines,
however notifications of the type C<debug> will only be passed in
development mode.


=head2 notifications

  %= notifications 'humane' => [qw/warn error success/];
  %= notifications 'html';

  $c->render(json => $c->notifications(json => {
    text => 'My message'
  }));

Serve notifications to your user based on an engine.
The engine's name has to be passed as the first parameter
and the engine has to be L<registered|/register> in advance.
Notifications won't be invoked in case no notifications are
in the queue and no further engine parameters are passed.
Engine parameters are documented in the respective plugins.

The engine's name will be camelized. If no namespace is given,
the default namespace is C<Mojolicious::Plugin::Notifications>.

In case no engine name is passed to the notifications method,
an L<assets object|Mojolicious::Plugin::Notifications::Assets>
is returned, bundling all registered engines' assets for use
in the L<AssetPack|Mojolicious::Plugin::AssetPack> pipeline.

  # Register Notifications plugin
  app->plugin('Notifications' => {
    Humane => {
      base_class => 'libnotify'
    },
    Alertify => 1
  });

  # Register AssetPack plugin
  app->plugin('AssetPack');

  # Add notification assets to pipeline
  app->asset('myApp.js'  => 'myscripts.coffee', app->notifications->scripts);
  app->asset('myApp.css' => 'mystyles.scss', app->notifications->styles);

  %# In templates embed assets ...
  %= asset 'myApp.js'
  %= asset 'myApp.css'

  %# ... and notifications (without assets)
  %= notifications 'humane', -no_include;

B<The asset helper option is experimental and may change without warnings!>


=head1 ENGINES

L<Mojolicious::Plugin::Notifications> bundles a couple of different
notification engines, but you can
L<easily write your own engine|Mojolicious::Plugin::Notifications::Engine>.


=head2 Bundled engines

The following engines are bundled with this plugin:
L<HTML|Mojolicious::Plugin::Notifications::HTML>,
L<JSON|Mojolicious::Plugin::Notifications::JSON>,
L<Humane.js|Mojolicious::Plugin::Notifications::Humane>, and
L<Alertify.js|Mojolicious::Plugin::Notifications::Alertify>,


=head1 HOOKS

  app->hook(
    before_notifications => sub {
      my ($c, $notes) = @_;
      $c->app->log('Served ' . $notes->size . ' notifications to ' . $c->stash('user'));
    });

This hook is emitted before any notifications are rendered.
The hook passes the current controller object and a L<Mojo::Collection>
object including all notes.
The hook is emitted no matter if notifications are pending.

B<This hook is EXPERIMENTAL!>


=head1 SEE ALSO

If you want to use C<Humane.js> without L<Mojolicious::Plugin::Notifications>,
you should have a look at L<Mojolicious::Plugin::Humane>,
which was the original inspiration for this plugin.

Without my knowledge (due to a lack of research by myself),
L<Mojolicious::Plugin::BootstrapAlerts> already established
a similar mechanism for notifications using Twitter Bootstrap
(not yet supported by this module).
Accidentally the helper names collide - I'm sorry for that!
On the other hands, that makes these modules in most occasions
compatible.


=head1 HINTS

As flash information is stored in the session, notifications may be lost
in case the session expires using C<session(expires =E<gt> 1)>.


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Notifications


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014-2018, L<Nils Diewald|https://nils-diewald.de/>.

Part of the code was written at the
L<Mojoconf 2014|http://www.mojoconf.org/mojo2014/> hackathon.

This program is free software, you can redistribute it
and/or modify it under the terms of the Artistic License version 2.0.

=cut
